cms/src/main/java/uk/nhs/digital/externalstorage/workflow/externalFilePublish/ExternalFilePublishTask.java

```diff
@@ -1,11 +1,119 @@
+import org.onehippo.repository.documentworkflow.DocumentHandle;
+import org.onehippo.repository.documentworkflow.DocumentVariant;
+import uk.nhs.digital.externalstorage.ExternalStorageConstants;
 import uk.nhs.digital.externalstorage.s3.S3Connector;
 import uk.nhs.digital.externalstorage.workflow.AbstractExternalFileTask;
+import uk.nhs.digital.ps.PublicationSystemConstants;
+
+import javax.jcr.Node;
+import javax.jcr.NodeIterator;
+import javax.jcr.RepositoryException;
 
 public class ExternalFilePublishTask extends AbstractExternalFileTask {
 
-    protected void setTargetStatus(S3Connector s3Connector, final String resource) {
-        s3Connector.publishResource(resource);
+    private static final String DOCUMENTS_ROOT_FOLDER = "/content/documents";
+
+    protected void setResourcePermission(S3Connector s3Connector, final NodeIterator resourceNodes) throws RepositoryException, WorkflowException {
+        Node variantNode = getVariant().getNode(getWorkflowContext().getInternalWorkflowSession());
+
+        if (isPublication(variantNode)) {
+            // if publishing publication we need to check "PUBLICLY ACCESSIBLE" flag, and also update all datasets
+            setResourcePermissionOnPublication(s3Connector, resourceNodes);
+        } else if (isDataset(variantNode)) {
+            // when publishing dataset, we need to check parent LIVE publication state
+            publishResources(s3Connector, resourceNodes, isParentPublicationFinalised(variantNode));
+        } else {
+            // no additional logic for other document types. Published == public attachments
+            publishResources(s3Connector, resourceNodes, true);
+        }
+    }
+
+    private void publishResources(S3Connector s3Connector, NodeIterator resourceNodes, boolean shouldBePublic) throws RepositoryException {
+        for (Node node; resourceNodes.hasNext(); ) {
+            node = resourceNodes.nextNode();
+            if (node.hasProperty(ExternalStorageConstants.PROPERTY_EXTERNAL_STORAGE_REFERENCE)) {
+                String externalResource = node
+                    .getProperty(ExternalStorageConstants.PROPERTY_EXTERNAL_STORAGE_REFERENCE)
+                    .getString();
+
+                if (shouldBePublic) {
+                    s3Connector.publishResource(externalResource);
+                } else {
+                    s3Connector.unpublishResource(externalResource);
+                }
+            }
+        }
+    }
+
+    private void setResourcePermissionOnPublication(S3Connector s3Connector, NodeIterator resourceNodes) throws RepositoryException {
+        Node variantNode = getVariant().getNode(getWorkflowContext().getInternalWorkflowSession());
+
+        publishResources(s3Connector, resourceNodes, isPublicationFinalised(variantNode));
+
+        // now find all datasets and publish resources
+        // only if publication name is "content"
+        if ("content".equals(variantNode.getName())) {
```

Comment:
`"content"` could be moved to `PublicationSystemConstants`?

Reply:
yes !!